polygeist/tools/cgeist/Lib/ValueCategory.cc

```diff
@@ -804,6 +805,30 @@ ValueCategory ValueCategory::FAdd(OpBuilder &Builder, Location Loc,
   return FPBinOp<arith::AddFOp>(Builder, Loc, val, RHS);
 }
 
+ValueCategory ValueCategory::FMA(OpBuilder &Builder, Location Loc,
+                                 ValueCategory Addend, bool NegMul,
+                                 bool NegAdd) const {
+  auto MulOp = val.getDefiningOp<arith::MulFOp>();
+
+  assert(MulOp && "Expecting arith.mul operation");
+
+  Value MulOp0 = MulOp.getLhs();
+  Value MulOp1 = MulOp.getRhs();
+  if (NegMul)
+    MulOp0 =
+        ValueCategory(MulOp0, /*IsReference=*/false).FNeg(Builder, Loc).val;
+  if (NegAdd)
+    Addend = Addend.FNeg(Builder, Loc);
+
+  ValueCategory FMulAdd(
+      Builder.create<math::FmaOp>(Loc, MulOp0, MulOp1, Addend.val),
+      /*IsReference=*/false);
+
+  MulOp.erase();
```

Comment:
Same question as above, should we rely on later passes to remove dead operations?